Query for competing orders

jonallured opened the pull request:
This PR exposes a new query for competing orders given an order id and seller id:

```gql
query($orderId: ID!) {
  competingOrders(orderId: $orderId) {
    totalCount
  }
}
```

There was some discussion over on #372 as to the approach for the consumer of this query, but I think this primitive lines up with whatever approach we take there.

In terms of the approach here, I've mostly copy/pasted from the orders query and then added a `Order#competing_orders` method to do the database query - open to ideas!

closes #372

sepans commented on app/graphql/types/query_type.rb:
```diff
@@ -25,6 +25,12 @@ class Types::QueryType < Types::BaseObject
+  field :competingOrders, Types::OrderConnectionWithTotalCountType, null: true, connection: true do
+    description 'Find list of competing orders'
+    argument :order_id, ID, required: true
+    argument :seller_id, String, required: true
```
Do we need the `sellerId`?

jonallured replied:
I think we need it in order to check that we're safe to return the results, but I only copy/pasted from the way that `validate_orders_request!` works. What do you think @ashkan18?

jonallured commented:
@sepans whoa, you've blown my mind!! I think you're right - just going to read through that query again with this use case in mind and might explore with a test or two, but really nice catch!! 🏆 ❤️ 💯

sepans commented on app/models/order.rb:
```diff
@@ -117,6 +117,12 @@ class Order < ApplicationRecord
+  def competing_orders
+    artwork_ids = line_items.pluck(:artwork_id)
+    conditions = 'orders.id != ? AND orders.state = ? AND line_items.artwork_id IN (?)'
+    Order.joins(:line_items).where(conditions, id, SUBMITTED, artwork_ids)
```
I don't exactly know how the edition sets are handled in exchange but is ☝️ query catch the competing orders on an edition set artwork?

jonallured replied:
Yeah you're right, this should be more like:

```
artwork_ids, edition_set_ids = line_items.pluck(:artwork_id, :edition_set_id)
conditions = 'orders.id != ? AND orders.state = ? AND (line_items.artwork_id IN (?) OR line_items.edition_set_id IN (?))'
Order.joins(:line_items).where(conditions, id, SUBMITTED, artwork_ids, edition_set_ids)
```

jonallured replied:
Ok, I've addressed this with bd32ff4.

jonallured commented:
Ok after much discussion with @sepans I'm back to thinking that adding this query makes sense. In my context, I've only got an order id, not an artwork id so the line items query that he mentioned isn't quite right.

ashkan18 commented on app/graphql/types/query_type.rb:
```diff
@@ -47,6 +53,12 @@ def line_items(args = {})
 
+  def competing_orders(params)
+    order = Order.find(params[:order_id])
+    validate_competing_orders_request!(params, order)
```
if we remove `seller_id` we can just use `validate_order_request`

jonallured replied:
Cool, good call - I really wasn't sure what to do here!! 😝 

I'll get a commit in that validates without passing in the seller id. 👍

ashkan18 commented on app/graphql/types/query_type.rb:
```diff
@@ -25,6 +25,12 @@ class Types::QueryType < Types::BaseObject
   end
 
+  field :competingOrders, Types::OrderConnectionWithTotalCountType, null: true, connection: true do
+    description 'Find list of competing orders'
```
should we give option to filter by `mode`? I can imagine `buy` orders have higher priority over `offer` ones.

jonallured replied:
I considered this, but I don't currently need it so I left it out. We can always add it later if need be!

ashkan18 commented on app/models/order.rb:
```diff
@@ -117,6 +117,17 @@ class Order < ApplicationRecord
     end
   end
 
+  def competing_orders
```
should we use scope for this? then we can also pass in `mode` as a param if needed and it will be more obvious that it can be chained.

jonallured replied:
I'm not a fan of scopes, I think they end up getting too hairy and I especially dislike them if they are complicated. I think a nice place for this business logic is a method on an order.

ashkan18 replied:
i was just thinking of case someone wanted to chain the result of this and add extra where clauses like `Order.competing_orders.where(mode: ?)` but we can cross that bridge when we get there :)

jonallured commented:
@ashkan18 yeah that's what @sepans and I were discussing - in his use case for Gravity hitting Exchange looking for line items he's blocking artwork deletion so he's got artwork/edition set ids, but in my use case, I'm drawing the order detail page and only have the order id. I think this approach makes more sense there.

jonallured commented:
Ok with 2520f5e I've switched to only passing in an order id and using the existing order validation based on either being trusted or the context of the query. I think I've addressed all feedback but lmk if I've missed something!! ❤️

ashkan18 commented on app/models/order.rb:
```diff
@@ -117,6 +117,17 @@ class Order < ApplicationRecord
+      AND orders.state = ?
+      AND (line_items.artwork_id IN (?) OR line_items.edition_set_id IN (?))
+    SQL
+    Order.joins(:line_items).where(conditions, id, SUBMITTED, artwork_ids, edition_set_ids)
```
I still think we need some sort of "sort" here, either having the default one (`created_at: :asc` ?)  or give mutation the option of passing sorts, what do you think?

jonallured commented on .rubocop.yml:
```diff
@@ -29,7 +29,7 @@ Metrics/MethodLength:
 
 Metrics/ClassLength:
-  Max: 190
+  Enabled: false
```
idk, maybe this is controversial, but it seems like just bumping this number up and up as I work is a waste of time - are people cool with me just disabling this sucker?

ashkan18 replied:
haha 😅 I think so, I generally agree with making classes, methods and etc short but i can't easily put a number on it!  I _think_ disabling this makes sense as long as we still look for opportunities to make shorter classes when possible 👍

jonallured commented:
Ok, I had a couple wacky commits there that I've force pushed to fix. I have added the default `order` clause in 4ac44e6 and then with 46c694f I regenerated the RuboCop todo file and removed the class length setting. My thinking there is that we should still have RuboCop complain about long classes but the current offenders have been listed in the todo file so that we can come back to them and clean up.

ashkan18 commented:
🙌 thanks! lets do this

yuki24 commented on app/models/order.rb:
```diff
@@ -117,6 +117,21 @@ class Order < ApplicationRecord
+      orders.id != ?
+      AND orders.state = ?
+      AND (line_items.artwork_id IN (?) OR line_items.edition_set_id IN (?))
+    SQL
```
Let's see if I can scope-ify this raw query...
